RetroPlayer: Cache/pre-compile output shaders

VelocityRa opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->
* Prevents stutering induced from the compilation of output shaders, when switching between them in the Video Filters OSD.

* Also fixes a bug where the shaders would compile continously in the Video Filters OSD.

I initially started working on an `RPWinOutputShaderManager` that would hold the output shaders and map them to scaling methods to make `RPWinRenderer` usage of shaders simpler, but it was getting a bit out of hand so I just did the logic in the renderer.

We won't have many other supported scaling types for this anyway, plus the output shader probably won't be used very much when [video shaders](https://github.com/garbear/xbmc/pull/86) are merged :)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
UI stuttering should be avoided when possible and in this case it was pretty trivial to do.

Also fixes the aforementioned bug. Sorry about that, I'll test stuff better next time.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
Tested by loading a game, going into the filters osd and switching between the 2 scaling options.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

VelocityRa committed:
RetroPlayer: Cache/pre-compile output shaders

Prevents stutering induced from the compilation of output shaders, when switching between them in the Video Filters OSD.

Commits by garbear:

Revert some logic

The reason we don't test renderers is because they might not be created
yet. Buffer pools, on the other hand, are created when RPRenderManager is
created, so we ask them about support. Buffer pools, in turn, query either
static functions of the renderer, or process info (which is also created
upon startup), or local properties of the buffer pool.

Revert some added logic

The default scaling method must be supported by at least one renderer, so
we don't need to check support here.

Revert changes to GUI texture renderer

Scaling method support is implicitly defined by the types of buffer pools
created. For example, on Windows, only linear scaling is supported, so
only a buffer pool that supports linear scaling is created.

Simplify shader construction

The lambda approach succeeds in reducing the amount of code, but adds
complexity. In this case, because it's only a few lines, it would be
better to simply duplicate the code to make the code much more
understandable. We're not writing code for the computer, we're writing
code for people. We want the code to be as straight-forward as possible,
and this goal trumps saving a few lines.

That comment withstanding, we could move handling of the default scaling
method to another place and greatly simplify this function.

Move output shaders to buffer pool

squashme

garbear commented on xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPWinRenderer.h:
```diff
@@ -123,12 +123,11 @@ namespace RETRO
-    std::unique_ptr<CRPWinOutputShader> m_outputShader;
-    ESCALINGMETHOD m_prevScalingMethod = VS_SCALINGMETHOD_AUTO;
+    std::unique_ptr<CRPWinOutputShader> m_outputShaderLinear;
+    std::unique_ptr<CRPWinOutputShader> m_outputShaderNearest;
```
How about a map from ESCALINGMETHOD to CRPWinOutputShader? That way we don't have to add a member variable for every scaling method.

VelocityRa replied:
Yeah I was trying to avoid that and go for the solution with less code, but I'll try it.

garbear replied:
More code is better than less code if it's scalable and maintainable

garbear commented on xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPWinRenderer.cpp:
```diff
@@ -269,19 +260,29 @@ bool CRPWinRenderer::SupportsScalingMethod(ESCALINGMETHOD method)
+        outputShader = m_outputShaderNearest.get();
+      else
+        // default to linear interpolation
+        outputShader = m_outputShaderLinear.get();
```
Might as well default to `CRPProcessInfo::GetDefaultScalingMethod()`

garbear commented:
@VelocityRa https://github.com/VelocityRa/xbmc/pull/25

garbear committed:
RetroPlayer: Fix black screen when shader fails to compile

This fixes a black screen shown when launching a game with a scaling
method not supported by any buffer pools.

garbear commented:
jenkins build this please
